---
title: "GSAdapt"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## 1. Brief introduction

GSAdapt, also known as Gene Set Adaptation, is a package designed for adapting gene sets, either published or created in-house, to analyze specific transcriptomic datasets. This functionality is particularly valuable when a gene set originates from a biological context different from the dataset under analysis. GSAdapt takes a SummarizedExperiment object and a dataframe containing feature information and their respective gene set names as input.

To load the GSAdapt package in RStudio, you can use the following command:

```R
library(GSAdapt)
```

GSAdapt filters features based on matrix expression row names and generates a correlation matrix for constructing a graph. Depending on user preferences, GSAdapt can be more permissive or more restrictive in its treatment of graph features. Additionally, users have the option to save the graph structure in their default directory to facilitate further exploration using various visualization methods.

## The gsadapt function

The `gsadapt` function is the main feature of the GSAdapt package. Users require two primary arguments:

1. A SummarizedExperiment object, which can be created from scratch using the `experiment_to_se` function. No specific information about genomic ranges is necessary.
2. A dataframe containing feature information. This dataframe consists of two columns: the gene symbol in the first column and the name of the gene set to be tested in the second column.

With these two arguments, `gsadapt` generates an igraph object and saves information about adapted genes and graph properties. However, users must specify three additional arguments:

a) `mggcc`: Median gene-to-gene correlation cutoff, a number between 0 and 1 indicating the threshold at which gene-to-gene correlations should be considered. This requirement determines which genes and connections are included in the graph.

b) `ncec`: Node centrality eigenvalue cutoff, a number between 0 and 1. A value of 0 represents genes without connections, while 1 indicates the most highly connected and central gene. By default, the gene with the highest number of neighbors and connections is assigned a value of 1. This cutoff allows users to adjust the centrality of genes within the graph.

c) `plot_graph`: A logical argument indicating whether the graph should be plotted and saved in a directory. Setting `plot_graph = TRUE` generates three graphs: the first includes all genes, even those without connections; the second removes isolated genes and provides information about the graph's structure; and the last graph applies the `ncec` cutoff to filter out isolated genes and those with low associations with the graph structure.

## Examples of Using GSAdapt

Here are basic examples of using `gsadapt` with the airway SummarizedExperiment dataset to create two gene sets:

```R
### Two or more signatures (permissive filter)
library(SummarizedExperiment)
library(airway)

data(airway, package = "airway")

Signature <- data.frame(Gene_ID = head(rownames(assay(airway)), n = 20), 
                        Gene_Set = c(rep("GS1", 10), rep("GS2", 10)))

DataMeta <- gsadapt(se = airway, 
                    gs_df = Signature,
                    mggcc = 0.2,
                    ncec = 0.1,
                    plot_graph = FALSE)

### Two or more signatures (restrictive filter)

data(airway, package = "airway")

Signature <- data.frame(Gene_ID = head(rownames(assay(airway)), n = 20), 
                        Gene_Set = c(rep("GS1", 10), rep("GS2", 10)))

DataMeta <- gsadapt(se = airway, 
                    gs_df = Signature,
                    mggcc = 0.6,
                    ncec = 0.3,
                    plot_graph = FALSE)
```

## Other GSAdapt Functions

1. `experiment_to_se`: This function generates a SummarizedExperiment object from an expression matrix and metadata containing sample information.

2. `Signature_input_to_GSAdapt`: This function allows users to format their gene sets for use with `gsadapt`.

